### PR TITLE
[ray] launch multiple GPU with ray

### PR DESCRIPTION
**Pull request by lihuahua123**
Support Ray to start the pipeline

### CODE REVIEW

File: xfuser/worker/worker.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+from xfuser import (
+    xFuserPixArtAlphaPipeline,
+    xFuserPixArtSigmaPipeline,
+    xFuserFluxPipeline,
+    xFuserStableDiffusion3Pipeline,
+    xFuserHunyuanDiTPipeline,
+)
+
+class WorkerBase(ABC):
+    """Worker interface that allows vLLM to cleanly separate implementations for
```
Comment (feifeibear) [inline xfuser/worker/worker.py]: if you copy code from vllm you should explicitly mention the source,

File: xfuser/worker/worker.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        raise NotImplementedError
+
+
+class Worker(WorkerBase):
+    """A worker class that executes (a partition of) the model on a GPU.
+
+    Each worker is associated with a single GPU. The worker is responsible for
+    maintaining the KV cache and executing the model on the GPU. In case of
+    distributed inference, each worker is assigned a partition of the model.
+    """
```
Comment (feifeibear) [inline xfuser/worker/worker.py]: carefully review the comments. adapt them to dit tasks

**Comment by feifeibear** [comment]
PR实现了通过ray方式启动多进程。参考vllm使用RayGPUExecutor来管理多个worker，每个worker执行diffusers pipefline的逻辑。

目前这种方式和torchrun启动程序（example.py)用法差别太大。

我建议设计一个DiffusionPipeline的Ray分布式版本，RayDiffusionPipeline，然后这个类提供from_pretrained，forward等接口。

PR中hardcode了一些地方，比如对模型初始化text_encoder处理，因为目前text_encoder是没有多卡切分的，可以让每个worker都重复载入text_encoder，希望尽量保持和torchrun接口一致性。

File: setup.py [hunk 3]
```diff
@@ -39,7 +39,8 @@ def get_cuda_version():
             "imageio",
             "imageio-ffmpeg",
             "optimum-quanto",
-            "flash_attn>=2.6.3"
+            "flash_attn>=2.7.0",  # flash_attn>=2.7.0 with torch>=2.4.0 wraps ops with torch.ops
```
Comment (feifeibear) [inline setup.py]: revert it back to 2.6.3

File: xfuser/config/args.py [hunk 4]
```diff
@@ -151,6 +154,17 @@ def add_cli_args(parser: FlexibleArgumentParser):
...
+        runtime_group.add_argument(
+            "--use_ray",
+            action="store_true",
+            help="Enable ray to run inference in multi-card",
+        )
+        parallel_group.add_argument(
+            "--ray_world_size",
+            type=int,
+            default=1,
+            help="World size.",
```
Comment (feifeibear) [inline xfuser/config/args.py]: the number of ray works (world_size for ray)

File: examples/run.sh [hunk 5]
```diff
@@ -3,7 +3,7 @@ set -x
 export PYTHONPATH=$PWD:$PYTHONPATH
 
 # Select the model type
-export MODEL_TYPE="Flux"
+export MODEL_TYPE="Sd3"
```
Comment (feifeibear) [inline examples/run.sh]: revert updates on this file

File: tests/executor/test_ray.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
+import time
+import os
+import torch
+import torch.distributed
+from transformers import T5EncoderModel
+from xfuser import xFuserStableDiffusion3Pipeline, xFuserArgs
+from xfuser.executor.gpu_executor import RayDiffusionPipeline
+from xfuser.config import FlexibleArgumentParser
+from xfuser.executor.gpu_executor import RayDiffusionPipeline
```
Comment (feifeibear) [inline tests/executor/test_ray.py]: remove this line

File: xfuser/executor/gpu_executor.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
+# Copyright 2024 The xDiT team.
```
Comment (feifeibear) [inline xfuser/executor/gpu_executor.py]: change the file to pipeline_utils.py ? refer to diffusers

File: xfuser/worker/worker.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
+
```
Comment (feifeibear) [inline xfuser/worker/worker.py]: put xfuser/worker -> xfuser/ray/worker ?

File: xfuser/executor/gpu_executor.py (hunk 7 above)
Comment (feifeibear) [reply xfuser/executor/gpu_executor.py]: put xfuser/executor -> xfuser/ray/pipeline ?

File: tests/executor/ray_run.sh [hunk 9]
```diff
@@ -0,0 +1,68 @@
+set -x
```
Comment (feifeibear) [inline tests/executor/ray_run.sh]: It is not appropriate to place a complete model inference script directly under the `tests` directory.

I suggest placing the Ray example scripts under the `examples/ray` directory.

**Review by feifeibear — CHANGES_REQUESTED** [review]

File: xfuser/executor/gpu_executor.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+from xfuser.model_executor.pipelines.base_pipeline import xFuserPipelineBaseWrapper
+logger = init_logger(__name__)
+
+
+class GPUExecutor(BaseExecutor):
+    def _init_executor(self):
+        pass
+
+
+class RayDiffusionPipeline(GPUExecutor):
```
Comment (feifeibear) [inline xfuser/executor/gpu_executor.py]: @Eigensystem do you think RayDiffusionPipeline should subclass DiffusionPipeline?

Comment (Eigensystem) [reply xfuser/executor/gpu_executor.py]: Maybe not. RayDiffusionPipeline can simply use `__getattr__` method to visit the actors' member. So if the actors derive from `DiffusionPipeline` or wrap a `DiffusionPipeline` object, it is unnecessary for RayDiffusionPipeline to inherit from DiffusionPipeline.

File: examples/ray/ray_flux_example.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    end_time = time.time()
+    elapsed_time = end_time - start_time
+
+    print(f"elapsed time:{elapsed_time}")
+    if not os.path.exists("results"):
+        os.mkdir("results")
+    # output is a list of results from each worker, we take the last one
+    for i, image in enumerate(output[-1].images):
+        image.save(
+            f"/data/results/{model_name}_result_{i}.png"
```
Comment (feifeibear) [inline examples/ray/ray_flux_example.py]: save to a relative path ./results/xxx

File: xfuser/config/config.py [hunk 12]
```diff
@@ -188,6 +192,9 @@ class ParallelConfig:
     sp_config: SequenceParallelConfig
     pp_config: PipeFusionParallelConfig
     tp_config: TensorParallelConfig
+    distributed_executor_backend: Optional[str] = None
+    world_size: int = 1 # FIXME: remove this
+    worker_cls: str = "xfuser.ray.worker.worker.Worker"
```
Comment (feifeibear) [inline xfuser/config/config.py:196]: do we need distributed_executor_backend and worker_cls?

**Review by feifeibear — CHANGES_REQUESTED** [review]

File: xfuser/config/config.py (hunk 12 above)
Comment (lihuahua123) [reply xfuser/config/config.py:196]: We don't need distributed_executor_backend, but we need worker_cls for ray to initial worker by its class name
```
def init_worker(self, *args, **kwargs):
      worker_class = resolve_obj_by_qualname(
          self.worker_cls)
      self.worker = worker_class(*args, **kwargs)
      assert self.worker is not None
```

**Review by feifeibear — APPROVED** [review]
LGTM

**Pull request merged**